### PR TITLE
Updates "Elm at Rakuten" post

### PR DESCRIPTION
**Pull request by lucamug**
Updates "Elm at Rakuten" post, also it renames the folder from "elm-in-rakuten" to "elm-at-rakuten"

### CODE REVIEW

File: content/post/elm-at-rakuten/index.md [hunk 1]
```diff
@@ -1,91 +1,100 @@
...
-Moreover, **The Elm Architecture**, included in the Elm language, seemed an excellent way to structure an application. It ended up being one of the most influential Elm ideas in the front-end world, more about this later.
+Moreover, **The Elm Architecture**, included in the Elm language, **seemed an excellent way to structure an application**. It ended up being one of the most influential Elm ideas in the front-end world, more about this later.
 
 ![The Elm Architecture](./images/the-elm-architecture.jpg)
 *The Elm Architecture, illustration by [Kolja Wilcke](https://twitter.com/01k), licensed under [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/).*
 
 So, we started learning Elm by building some prototypes, and the first results were promising. But the technology stack was still heavily server-side based, mostly PHP, and the adoption was proceeding slowly.
 
-### 🕒　One year later
+### 🕒　One year later
```
Comment (jcayzac) [inline content/post/elm-at-rakuten/index.md:51]: trailing space

File: content/post/elm-at-rakuten/index.md [hunk 2]
```diff
@@ -1,91 +1,100 @@
...
 
-Elm was promising that all of our functions would be pure by design and that there would be **no errors at runtime**[^errors-at-runtime] anymore.
+Elm was promising that **all of our functions would be pure** by design and that there would be **no errors at runtime**[^errors-at-runtime] anymore.
 
-[^errors-at-runtime]: In the front-end, **errors at runtime** are errors that happen in the browser. These errors can completely halt the functionality of the website and you, as the creator, may not even know about them because they are happening on other peoples' devices. Some tools allow you to get notified when these errors happen. These errors are usually JavaScript errors, for example, trying to access values that are null or undefined.
+[^errors-at-runtime]: In the front-end, **errors at runtime** are errors that happen in the browser. These errors can completely halt the functionality of the website and you, as the creator, may not even know about them because they are happening on other peoples' devices. Some tools allow you to get notified when these errors happen. These errors are usually JavaScript errors, for example, trying to access values that are null or undefined.
```
Comment (jcayzac) [inline content/post/elm-at-rakuten/index.md:42]: trailing space

File: content/post/elm-at-rakuten/index.md [hunk 3]
```diff
@@ -1,91 +1,100 @@
...
-1. [Elm as an influencer](#10-elm-as-an-influencer)
-1. [The Elm Architecture](#11-the-elm-architecture)
-1. [The Elm debugger](#12-the-elm-debugger)
-1. [Elm-UI, the alternative to CSS/HTML](#13-elm-ui-the-alternative-to-csshtml)
-1. [Readability and Elm syntax](#14-readability-and-elm-syntax)
-1. [Refactoring](#15-refactoring)
-1. [Hiring](#16-hiring)
-1. [Fast performance and small assets](#17-fast-performance-and-small-assets)
+1. [Guarantees](#1-guarantees)
+1. [Controlled state](#2-controlled-state)
```
Comment (jcayzac) [inline content/post/elm-at-rakuten/index.md:80]: All the numbers are `1`.

**Comment by jcayzac** [comment]
Maybe pass the .md through prettier?

File: content/post/elm-at-rakuten/index.md (hunk 1 above)
Comment (lucamug) [reply content/post/elm-at-rakuten/index.md]: prettified

File: content/post/elm-at-rakuten/index.md (hunk 2 above)
Comment (lucamug) [reply content/post/elm-at-rakuten/index.md]: prettified

File: content/post/elm-at-rakuten/index.md (hunk 3 above)
Comment (lucamug) [reply content/post/elm-at-rakuten/index.md:77]: I believe this is ok

**Comment by jcayzac** [comment]
@lucamug What is the canonical URL?

1. <https://dev.to/lucamug/elm-6m8>?
2. <https://engineering.rakuten.today/post/elm-at-rakuten>?

If **1**, you're missing the following in your frontmatter:

```yaml
seo:
  canonical: https://dev.to/lucamug/elm-6m8
```

If **2**, you'll need to edit your post at <https://dev.to/lucamug/elm-6m8> and set the canonical url to https://engineering.rakuten.today/post/elm-at-rakuten once it's up (dev.to lets you set the canonical url).

File: content/post/elm-at-rakuten/index.md [hunk 4]
```diff
@@ -599,74 +737,105 @@ b = a + 2
...
+[Index](#index) | [Top](#top)
+
+---
+
+## 17. Hiring
+
+Hiring for a language that is not mainstream has some downsides. For example, not many developers probably know it fluently.
 
-Hiring for a language that is not mainstream has some downsides. For example, not many developers are already fluent in it.
+However, the feeling, from engaging in the Elm slack channel, is that there are probably now more developers that know Elm compared to the number of available Elm job positions, making it possible to find developers that already master it.
```
Comment (jcayzac) [inline content/post/elm-at-rakuten/index.md:768]: It feels a bit like what somebody would say about COBOL (lots of starving devs and no jobs). Is that something we love about Elm?

**Comment by jcayzac** [comment]
When pasted into Grammarly, it reports a lot of typos and grammatical mistakes.

File: content/post/elm-at-rakuten/index.md (hunk 4 above)
Comment (lucamug) [reply content/post/elm-at-rakuten/index.md]: I removed this paragraph

**Comment by lucamug** [comment]
> @lucamug Duplicate content is very bad for SEO. What is the canonical URL?
> 
> 1. https://dev.to/lucamug/elm-6m8?
> 2. https://engineering.rakuten.today/post/elm-at-rakuten?
> 
> If **1**, you're missing the following in your frontmatter:
> 
> ```yaml
> seo:
>   canonical: https://dev.to/lucamug/elm-6m8
> ```
> 
> If **2**, you'll need to edit your post at https://dev.to/lucamug/elm-6m8 and set the canonical url to https://engineering.rakuten.today/post/elm-at-rakuten once it's up (dev.to lets you set the canonical url).

I went with **1** for the moment. I can change to **2** once the post is published, if necessary

**Comment by lucamug** [comment]
> When pasted into Grammarly, it reports a lot of typos and grammatical mistakes.

I fixed all the typos and grammatical mistakes that Grammarly reported

**Comment by jcayzac** [comment]
@lucamug 
> I went with **1** for the moment. I can change to **2** once the post is published, if necessary

So you're missing the `seo.canonical` thing in the frontmatter:

> > If **1**, you're missing the following in your frontmatter:
> > ```yaml
> > seo:
> >   canonical: https://dev.to/lucamug/elm-6m8
> > ```

**Comment by lucamug** [comment]
> So you're missing the `seo.canonical` thing in the frontmatter:

Is this not correct? https://github.com/rakutentech/engineering-blog/pull/10/commits/8fbe39c491b76b599cc135a6dee19cab390f9824

**Comment by jcayzac** [comment]
> > So you're missing the `seo.canonical` thing in the frontmatter:
> 
> Is this not correct? [8fbe39c](https://github.com/rakutentech/engineering-blog/commit/8fbe39c491b76b599cc135a6dee19cab390f9824)

Sorry, it wasn't showing in the diff when I tried. Now I can see it 👍🏼

**Review by jcayzac — APPROVED** [review]

**Comment by NemoOudeis** [comment]
`debugger.gif` is unused

File: content/post/elm-at-rakuten/index.md [hunk 5]
```diff
@@ -301,43 +350,56 @@ The main advantage of currying is the increased flexibility in combining functio
...
-## 8. Learnability
+[Index](#index) | [Top](#top)
 
-Elm is beginner-friendly. It doesn't mean that Elm is not sophisticated enough. It means that it is well designed. There are both simple constructs for beginners and complex constructs for masters. Complexity gets introduced gradually.[^introduced-gradually]
+---
 
-[^introduced-gradually]: For example *sandbox*, *element*, *document*, and *application* in [Elm-Browser](https://package.elm-lang.org/packages/elm/browser/latest/Browser); *get*, *post*, and *request* in [Elm-HTTP](https://package.elm-lang.org/packages/elm/http/latest/Http); *picture*, *animation*, and *game* in [Elm-Playground](https://package.elm-lang.org/packages/evancz/elm-playground/latest/Playground); etc.
+## 9. Learnability
+
+Elm is beginner-friendly. It doesn't mean that Elm is not sophisticated enough. It means that it is well designed. There are both simple constructs for beginners and complex constructs for masters. Complexity gets introduced gradually.[^introduced-gradually] This concept is sometimes called "gradual learning" or "progressive disclosure of complexity".[^progressive-disclosure-of-complexity]
```
Comment (NemoOudeis) [inline content/post/elm-at-rakuten/index.md:381]:
```suggestion
Elm is beginner-friendly. It doesn't mean that Elm is not sophisticated. It means that it is well designed. There are both simple constructs for beginners and complex constructs for masters. Complexity gets introduced gradually.[^introduced-gradually] This concept is sometimes called "gradual learning" or "progressive disclosure of complexity".[^progressive-disclosure-of-complexity]
```

File: content/post/elm-at-rakuten/index.md [hunk 6]
```diff
@@ -347,193 +409,241 @@ This feature is so crucial that it inspired a new coding style: **Compiler Drive
...
+[^representation-of-unidirectional-flow]: The illustration **A simple representation of the Elm Architecture** is from the [Elm Guide](https://guide.elm-lang.org/architecture/) used under the [CC BY-NC-ND 4.0](https://creativecommons.org/licenses/by-nc-nd/4.0/) [license](https://github.com/evancz/guide.elm-lang.org/blob/master/LICENSE).
 
 There are three building blocks in The Elm Architecture:
 
-* `Model` --- the state of your application, the only thing that can mutate
-* `view` --- a way to turn your state into HTML
-* `update` --- a way to update your state based on the `Model` and the *messages*
+- `Model` --- the state of your application, the only thing that can mutate
+- `view` --- a way to turn your state into HTML
+- `update` --- a way to update your state based on the `Model` and the _messages_
```
Comment (NemoOudeis) [inline content/post/elm-at-rakuten/index.md:487]: nitpick: should they all be capitalized? Like

```
- `Model`
- `View`
- `Update`
```

Or is there a reason to only capitalize model?

File: content/post/elm-at-rakuten/index.md [hunk 7]
```diff
@@ -599,74 +739,103 @@ b = a + 2
...
+[Index](#index) | [Top](#top)
 
-Hiring for a language that is not mainstream has some downsides. For example, not many developers are already fluent in it.
+---
 
-However, there are probably now more developers that know Elm compared to the number of available Elm job positions, making it easy to find Elm developers.
+## 17. Hiring
 
-Besides, expert Elm engineers probably are passionate about it and presumably have a broad and deep grasp of other technologies as well.
+Hiring for a language that is not mainstream has some downsides. For example, not many developers probably know it fluently.
```
Comment (NemoOudeis) [inline content/post/elm-at-rakuten/index.md:768]:
```suggestion
Hiring for a language that is not mainstream has some downsides. For example, not many developers know it fluently.
```

File: content/post/elm-at-rakuten/index.md [hunk 8]
```diff
@@ -701,30 +882,28 @@ Some consider this to be a benefit as it brings you out of your comfort zone and
...
 
 [^interop]: Elm provides several methodologies to communicate with JavaScript. Here is [an introduction to JavaScript interoperability](https://guide.elm-lang.org/interop/) with some [examples](https://github.com/elm-community/js-integration-examples).
 
+[Index](#index) | [Top](#top)
+
+---
+
 # Conclusion
 
 We mentioned some of the benefits of coding with Elm, or with a purely functional language for that matter. We also talked about the main issues.
```
Comment (NemoOudeis) [inline content/post/elm-at-rakuten/index.md:909]: On the line below (can't comment because not changed in this PR)

```diff
- The benefits are overwhelmingly in the majority, compared to the issues, and this is why we are happy with the choice we made.
+ For us the benefits are overwhelming, compared to the issues, and this is why we are happy with the choice we made.
```

File: content/post/elm-at-rakuten/index.md [hunk 9]
```diff
@@ -1,91 +1,99 @@
...
+title: "Elm at Rakuten"
+date: 2021-07-13T00:00:00+09:00
+hero: /post/elm-at-rakuten/images/elm-at-rakuten.png
 excerpt: "Our story of adopting the Elm language in Rakuten. The lessons we learned, and our likes and dislikes."
 timeToRead: 5
 authors:
   - Luca Mugnaini
 draft: false
+seo:
+  canonical: https://dev.to/lucamug/elm-6m8
```
Comment (NemoOudeis) [inline content/post/elm-at-rakuten/index.md:11]: ...once we should change this before we publish...

File: content/post/elm-at-rakuten/index.md [hunk 10]
```diff
@@ -95,82 +103,109 @@ In no particular order.
...
+[Index](#index) | [Top](#top)
 
-## 3. “Making impossible states impossible”
+---
+
+## 4. “Making impossible states impossible”
 
 ![Penrose Triangle](./images/penrose-triangle.jpg)
-*Penrose Triangle.*[^penrose-triangle]
+_Penrose Triangle.[^penrose-triangle]_
```
Comment (jcayzac) [inline content/post/elm-at-rakuten/index.md:208]: The Penrose triangle illustration has no license information.

File: content/post/elm-at-rakuten/index.md [hunk 11]
```diff
@@ -347,193 +409,241 @@ This feature is so crucial that it inspired a new coding style: **Compiler Drive
...
+
+[^variable-shadowing]: An example of **Elm making “best practices” the default**, is about **variable shadowing** (same variable name defined twice in an ambiguous way). While most linters produce warnings in case of variable shadowing, the Elm compiler generates an error and stops the compilation, until the variable shadowing issue is removed. More on this in the [Variable Shadowing](https://github.com/elm/compiler/blob/master/hints/shadowing.md) article.
+
+The compiler guarantees that all edge cases are covered, something difficult to achieve with hand-made unit tests. Another advantage of the compiler static analysis is that it is extremely fast and can provide the exact location of errors.[^elm-review]
 
-The compiler guarantees that all edge cases are covered, something difficult to achieve with hand-made unit tests. Another advantage of the compiler static analysis is that it is extremely fast and can provide the exact location of errors.
+[^elm-review]: To push the static analysis further, other tools can be used, like [Elm-Review](https://package.elm-lang.org/packages/jfmengels/elm-review/latest/).
 
 ![John Carmack on Elm Errors](./images/john-carmack-on-errors.png)
```
Comment (jcayzac) [inline content/post/elm-at-rakuten/index.md:429]: The Carmack tweet doesn't link to Twitter.

File: content/post/elm-at-rakuten/index.md [hunk 12]
```diff
@@ -347,193 +409,241 @@ This feature is so crucial that it inspired a new coding style: **Compiler Drive
...
 
 ![The Elm Architecture, animated](./images/the-elm-architecture.svg)
-*A simple representation of the unidirectional data flows in The Elm Architecture. (Source: The Elm Guide)*[^representation-of-unidirectional-flow]
+_A simple representation of the unidirectional data flows in The Elm Architecture. (Source: The Elm Guide) [^representation-of-unidirectional-flow]_
 
-[^representation-of-unidirectional-flow]: The illustration **A simple representation of the Elm Architecture** is from the as depicted in the [Elm Guide](https://guide.elm-lang.org/architecture/) used under the [CC BY-NC-ND 4.0](https://creativecommons.org/licenses/by-nc-nd/4.0/) [license](https://github.com/evancz/guide.elm-lang.org/blob/master/LICENSE).
+[^representation-of-unidirectional-flow]: The illustration **A simple representation of the Elm Architecture** is from the [Elm Guide](https://guide.elm-lang.org/architecture/) used under the [CC BY-NC-ND 4.0](https://creativecommons.org/licenses/by-nc-nd/4.0/) [license](https://github.com/evancz/guide.elm-lang.org/blob/master/LICENSE).
```
Comment (jcayzac) [inline content/post/elm-at-rakuten/index.md:481]: The data flow illustration links to the Elm guide but doesn't have the license information in the caption (you need to click on the link), which is not consistent with other illustrations.

**Comment by NemoOudeis** [comment]
@lucamug do we have a PR review ticket for this?

File: content/post/elm-at-rakuten/index.md (hunk 6 above)
Comment (lucamug) [reply content/post/elm-at-rakuten/index.md:487]: Model is a type and in Elm types are capitalized. view and update are functions and functions are not capitalized.

A familiar notation for Elm devs but probably awkward for non-Elm devs.

File: content/post/elm-at-rakuten/index.md (hunk 11 above)
Comment (lucamug) [reply content/post/elm-at-rakuten/index.md:429]: Fixed

File: content/post/elm-at-rakuten/index.md (hunk 10 above)
Comment (lucamug) [reply content/post/elm-at-rakuten/index.md:208]: This is original, made by me

File: content/post/elm-at-rakuten/index.md (hunk 9 above)
Comment (lucamug) [reply content/post/elm-at-rakuten/index.md:11]: Should I change now?

Comment (NemoOudeis) [reply content/post/elm-at-rakuten/index.md:11]: yeah let's do it!

**Review by NemoOudeis — APPROVED** [review]

File: content/post/elm-at-rakuten/index.md [hunk 13]
```diff
@@ -701,42 +882,59 @@ Some consider this to be a benefit as it brings you out of your comfort zone and
...
 Elm provides three ways to interact with external libraries: Flags, Ports, and Custom Elements.[^interop] All of them require you to write some JavaScript.
 
 In our case, for example, we must use a JavaScript library for handling payments.
 
-The required CSS, while using **Elm-UI**, is limited. In our applications, we have small snippets of CSS that are mainly tricks to support IE11.
+The required CSS, while using the library _Elm-UI_, is limited. In our applications, we have small snippets of CSS that are mainly tricks to support IE11.
 
 Related to this, Elm is probably not a good fit for short projects that require lots of integration with third-party JavaScript libraries.
 
 [^interop]: Elm provides several methodologies to communicate with JavaScript. Here is [an introduction to JavaScript interoperability](https://guide.elm-lang.org/interop/) with some [examples](https://github.com/elm-community/js-integration-examples).
```
Comment (boazy) [inline content/post/elm-at-rakuten/index.md:902]: I would put a `6. Limited Polymorphic Abstraction Power`. At least this is a personal issue for me.
To put it shortly, in JavaScript you would use (or abuse) duck typing (i.e. dynamic polymorphism) to do many different things. A lot of these things are better served by the simple type-safe facilities Elm offers like sum types and first order generics. But some of these things cannot be done with the type-safe abstraction Elm offers.

I think the easiest example (and something I ran into when working on Flipper) is a multi-page app. My memory might be a bit off, but each page had its own TEA and therefore its own model, view and update. The main TEA is responsible for routing and converting all the messages, models and commands between the pages. The problem is that this ends up in adding a lot of boilerplate for any page you add in multiple different places. It is quite easy to forget to add something and your new page will not work (it is quite hard to make a mistake in all the boring boilerplate code — the Elm type-system will generally protect you against that).
I could shave off some of this boilerplate by using _explicit type-classes_ (basically the approach detailed in [Scrap Your Type Classes](https://www.haskellforall.com/2012/05/scrap-your-type-classes.html)). This improves the situation a little bit, but still requires boilerplate.

We could have saved some of the boilerplate if Elm had a concept of type classes, or even just plain interfaces (as in Java). Another source of boilerplate (dealing with different model types and message types and converting between them) is something that could have been reduced by Higher-kinded Types or Associated Types.

The rebuttal for all of the above (and why I would still use Elm in many projects) is that the JS _does allow_ for more abstraction power, but at the same time it doesn't provide any standardized method or system of doing these abstraction. With JS we almost always end up with a big hard-to-understand tangle of abstractions created by different developers who had their own idea about the right abstraction. TypeScript gives a more type-safe way to define these abstractions, but doesn't alleviate the issue.

I keep thinking Elm is following the celebrated Go philosophy much better than Go. The much-hyped Go language design philosophy is ostensibly "Keep the language simple and you'll prevent developers from using the wrong abstraction and creating hard-to-maintain code". The problem is that Go has many complex features (magic built-in types like channels, slices, arrays and maps, multiple returns, the way channels work) that should have been part of the library. Go also forces you to make all your data mutable and makes closures extremely cumbersome to use (on purpose), so you end up with something that resembles the Algol of the 1960s or the Pascal of the 1970s. I thought this philosophy is a sham, until I started using Elm and for the first time in my life I was fine with trading some abstraction power for simplicity. I'm still not 100% sold on this approach, but I can understand the benefits.

I'm not sure how to phrase all of that nicely, especially since my Elm is a little rusty. But just my thoughts on what could perhaps be a future article.

**Review by boazy — CHANGES_REQUESTED** [review]
Small changes

File: content/post/elm-at-rakuten/index.md (hunk 13 above)
Comment (boazy) [reply content/post/elm-at-rakuten/index.md:902]: @lucamug Please treat this as a comment since it will be too big to write... You can just resolve this thing :)

Comment (lucamug) [reply content/post/elm-at-rakuten/index.md:902]: @boazy , thank you for the feedback.

The main issue mentioned about Flipper I think it comes from inexperience (it was one of our first Elm project) and lack of tooling. 

About the boilerplate, at the moment is quite standard to use the library `elm-spa` (https://www.elm-spa.dev/) to build Single Page Applications. The library that take care of most of the wiring and allows us to decided if a page need the TEA or is just a view.

The topic of typeclasses has been debated a lot in the Elm community (for example https://discourse.elm-lang.org/search?q=typeclasses) and it is still unclear to me how much they would improve the language. I think it could be a topic for another post.

File: content/post/elm-at-rakuten/index.md (hunk 9 above)
Comment (lucamug) [reply content/post/elm-at-rakuten/index.md]: Done

File: content/post/elm-at-rakuten/index.md [hunk 14]
```diff
@@ -95,82 +103,109 @@ In no particular order.
...
 # What we like about Elm
 
-## 1. Controlled state
+---
+
+## 1. Guarantees
+
+These are probably the most objective and important guarantees that Elm provides and that are difficult (impossible?) to find in other frameworks.
+
+- ⛔ No runtime exception.
```
[the remark this answers is not in the document]
Comment (lucamug) [reply content/post/elm-at-rakuten/index.md]: Changed

**Review by boazy — APPROVED** [review]
